### ISSUE TITLE
cocoapods cordova build ios fails with linker issues (works pefectly in Xcode UI)

### ISSUE THREAD

**Issue by jpgilchrist** [apache/cordova-ios#394]
I have been debugging and investigating this issue for almost three days. I am at my wits end and hopefully I have enough information now so you can give me an appropriate answer or we can identify an actual. 

First and foremost, I started with SO to do a bunch of research before I came here and posted my own question: https://stackoverflow.com/questions/51772407/cordova-cant-build-from-cli-but-build-from-xcode-xcworkspace-works-perfectly 

Here's what I'm finding:

`cordova build ios` fails with the following

```
/Users/jpgilchrist/Development/github/soschat/soschat-mobile-3.0/VeociSTAGE/platforms/ios/Pods/FirebaseCore/Firebase/Core/FIRNetwork.m:23:9: fatal error: 
      'GoogleToolboxForMac/GTMNSData+zlib.h' file not found
#import <GoogleToolboxForMac/GTMNSData+zlib.h>
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

```
** BUILD FAILED **


The following build commands failed:
	CompileC /Users/jpgilchrist/Library/Developer/Xcode/DerivedData/VeociSTAGE-awzedsneagcsyzbvjuozwzgwilon/Build/Intermediates.noindex/Pods.build/Debug-iphonesimulator/FirebaseCore.build/Objects-normal/x86_64/FIRNetwork.o FirebaseCore/Firebase/Core/FIRNetwork.m normal x86_64 objective-c com.apple.compilers.llvm.clang.1_0.compiler
```

However, if you actually open the .xcworkspace and then build/run from Xcode it works perfectly! This got me to thinking and a started down a deep rabbit hole. Eventually I tried `xcodebuild -workspace VeociSTAGE.xcworkspace -scheme VeociSTAGE` and it worked there as well!

This started me down a very deep rabbit hole learning way more about the innards of the cordova build process then I really needed to know (but, hey knowledge is knowledge and I'm happy I know it now). 

I found `platforms/ios/cordova/Api.js` and eventually the `xcodebuildArgs` and I deduced that a custom xcconfig is being used and I have no idea what in the xcconfig is causing it, but running it with the xcconfig causes my issues. Without it builds.

**Comment by jpgilchrist** [comment]
Also weird side note, I had to add a run_pods.sh script to get things to work as well... 

```
        <hook src="run_pods.sh" type="after_platform_add" />
```

**Comment by shazron** [comment]
hi @jpgilchrist, thanks for the report. Is there a plugin that we can use to repro the issue and investigate? So that we will be on the same page. Thanks!

**Comment by shazron** [comment]
Confirmed it with plugin cordova-plugin-firebase-messaging: 
```
/Users/shaz/Desktop/foo/platforms/ios/Pods/FirebaseCore/Firebase/Core/FIRLogger.m:18:9: fatal error:
      'GoogleUtilities/GULAppEnvironmentUtil.h' file not found
#import <GoogleUtilities/GULAppEnvironmentUtil.h>
```

Doesn't build with `cordova build` but builds properly with the workspace.

**Comment by shazron** [comment]
In the command line build:
```
    FRAMEWORK_SEARCH_PATHS = $(inherited) "${PODS_ROOT}/FirebaseAnalytics/Frameworks" "${PODS_ROOT}/FirebaseInstanceID/Frameworks" "${PODS_ROOT}/GoogleAppMeasurement/Frameworks"
```
 I manually checked the paths referenced, and I couldn't find `GoogleUtilities.framework` in there, so that is partially the root of the problem.

**Comment by shazron** [comment]
```
HEADER_SEARCH_PATHS = $(inherited) ${PODS_ROOT}/Firebase/CoreOnly/Sources "${PODS_ROOT}/Headers/Public" "${PODS_ROOT}/Headers/Public/Firebase" "${PODS_ROOT}/Headers/Public/FirebaseAnalytics" "${PODS_ROOT}/Headers/Public/FirebaseCore" "${PODS_ROOT}/Headers/Public/FirebaseInstanceID" "${PODS_ROOT}/Headers/Public/FirebaseMessaging" "${PODS_ROOT}/Headers/Public/GoogleUtilities" "${PODS_ROOT}/Headers/Public/Protobuf" "${PODS_ROOT}/Headers/Public/nanopb"
```

Note that the search paths are all for `Public` headers, when `<GoogleUtilities/GULAppEnvironmentUtil.h>` is a `Private` header.

**Comment by shazron** [comment]
The right `Headers` folder is sent, even the `Private` one, when the file is compiled. However, the path is wrong(!), it duplicates `Pods`. i.e. `-I/Users/shaz/Desktop/foo/platforms/ios/Pods/Pods/Headers/Private`

**Comment by shazron** [comment]
The duplicate `Pods` is the only difference I can see, when viewing the log for compiling the same file (where the error occurs) on the command line vs Xcode.

Found another bug: when I switched command line tools to Xcode 10 beta 4, it seemed to build (no errors) but I don't think it actually built anything (no build artifacts) and may have failed silently because of a config issue.

**Comment by shazron** [comment]
Found this duplicate issue: https://issues.apache.org/jira/browse/CB-14057

**Comment by jonathanrevell** [comment]
I'm also seeing these same issues with Cordova 8.0.0, XCode 9.4.1, Pod version 1.5.3. Works fine from inside the XCode UI. Fails no matter what I do when trying to run it via cordova build, etc.

Been stuck on this issue for several days now. Is there any known workaround?

**Comment by jonathanrevell** [comment]
I was able to find a workaround which involves bypassing cordova for now:
Reconstruct the command being passed to xcodebuild from cordova, you can see the list of arguments passed in at the bottom of the stack trace, but remove the -xcconfig argument altogether. (Based on this [comment](https://issues.apache.org/jira/browse/CB-14057?focusedCommentId=16487056&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16487056))

Your command should look something like this:
`xcodebuild clean build -workspace YourProjectName.xcworkspace -scheme YourProjectName -configuration Debug -destination generic/platform=iOS -archivePath YourProjectName.xcarchive CONFIGURATION_BUILD_DIR=.../cordova/platforms/ios/build/device SHARED_PRECOMPS_DIR=.../cordova/platforms/ios/build/sharedpch EMBEDDED_CONTENT_CONTAINS_SWIFT=YES ALWAYS_EMBED_SWIFT_STANDARD_LIBRARIES=NO LD_RUNPATH_SEARCH_PATHS="@executable_path/Frameworks"`

Replace YourProjectName with your project name, and note that the ... paths should be your actual paths

**Comment by dpogue** [comment]
I believe @erisu had run into a similar build issue where removing `-xcconfig` fixed the problem, but I thought he was testing against Xcode beta and I don't know if his attempt involved CocoaPods. He was trying to make a minimal reproducible test case to investigate further, so CC'ing him in case this issue provides some additional details.

**Comment by jpgilchrist** [comment]
@jonathanrevell yes I have deduced the same thing and @dpogue is correct it's something to do with the -xcconfig but no one really knows what. The trouble I'm finding is that no one seems to know if it's a crodova issue or an xcodebuild issue, but it seems people are learning towards an xcodebuild issue. 

I too am still waiting on a decent solution. I really don't want to change all of my automated build scripts because of this issue.

**Comment by jonathanrevell** [comment]
Lucky me I'm in the process of writing my build scripts for the first time and was tearing my hair out wondering what I did wrong with them only to find out there's some sort of cordova-xcodebuild bug right now. So I'm just going to be using xcodebuild straight up for the time being until the root cause is diagnosed and fixed

**Comment by erisu** [comment]
@dpogue The beta issue was something different.

I also seen the CocoaPods issue for a while now, in one of my environments. In that environment, I completely removed the xcconfig flag so that the iOS with CocoaPods apps will pass during build.

Since the xcconfig is applied for entire project's workspace, during code signing step, it tries to signing the libraries that were imported by CocoaPods. This caused the build to error.

**Comment by BobCaSUaL** [comment]
Step to workaround:

1. your need to add a exportOptions.plist used in step 3 and filled with
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
    <key>method</key>
    <string>YOUR_EXPORT_METHOD</string>
    <key>teamID</key>
    <string>YOUR_TEAM_ID_TAKEN_FROM_PROVISIONING</string>
    <key>provisioningProfiles</key>
    <dict>
        <key>YOUR_APP_ID</key>
        <string>YOUR_APP_UUID_TAKEN_FROM_PROVISIONING</string>
    </dict>
</dict>
</plist>
```
refer here https://devcenter.bitrise.io/tips-and-tricks/xcodebuild-export-options/ for more information.

2. Archive the project
xcodebuild clean build -workspace ./platforms/ios/YOUR_PROJECT_NAME.xcworkspace -scheme YOUR_PROJECT_NAME -configuration YOUR_CONFIGURATION -destination generic/platform=iOS archive -archivePath ./platforms/ios/build/YOUR_PROJECT_NAME.xcarchive

3. Generate the ipa:
xcodebuild -exportArchive -archivePath ./platforms/ios/build/YOUR_PROJECT_NAME.xcarchive -exportOptionsPlist YOUR_EXPORT_OPTIONS_PLIST_PATH -exportPath ./platforms/ios/build/device

YOUR_CONFIGURATION is the xcode configuration name, usually "Debug" and "Release".



But the real solution requires bugfix.

**Comment by erisu** [comment]
@jpgilchrist I just submitted a PR but will finalize tomorrow.

What I mean by `removing`, as I said previously, I only commented out the `xcconfig` option that is declared in the `build.js`

You can see the changes here: https://github.com/apache/cordova-ios/pull/398

**Comment by jpgilchrist** [comment]
@erisu looks like your PR was merged 👍

**Comment by sek** [comment]
> @erisu looks like your PR was merged 👍

I've confirmed this fixes the problem for me as well.  It's merged to master, but not 4.5.x branch.  Do we know what the release schedule is for the next ios platform version?
If 4.5.6 - I think this should be merged to that branch also.
If 4.6 with a new branch from master, then it will be included without additional merging.

In the meantime:

- I've downloaded the raw file as ios-updated-build.js in root of my project
- verified it's the only diff from the current default 4.5.5 ios version
- added a resource-file line to my config.xml under ios platform and that's got my local project building again
```
<platform name="ios">
     <resource-file src="ios-updated-build.js" target="../../cordova/lib/build.js" />
     ...
```

**Comment by mohamedsharaf** [comment]
> > @erisu looks like your PR was merged 👍
> 
> I've confirmed this fixes the problem for me as well. It's merged to master, but not 4.5.x branch. Do we know what the release schedule is for the next ios platform version?
> If 4.5.6 - I think this should be merged to that branch also.
> If 4.6 with a new branch from master, then it will be included without additional merging.
> 
> In the meantime:
> 
> * I've downloaded the raw file as ios-updated-build.js in root of my project
> * verified it's the only diff from the current default 4.5.5 ios version
> * added a resource-file line to my config.xml under ios platform and that's got my local project building again
> 
> ```
> <platform name="ios">
>      <resource-file src="ios-updated-build.js" target="../../cordova/lib/build.js" />
>      ...
> ```

this may help other this is the file to download
https://raw.githubusercontent.com/erisu/cordova-ios/4dd0586ebdf1650ae9079c98cae4d68b96bc08fe/bin/templates/scripts/cordova/lib/build.js

**Comment by erisu** [comment]
https://github.com/apache/cordova-ios/pull/398 has been released in `cordova-ios@5.0.0`.
